crates/view/src/planner.rs

```diff
@@ -79,12 +83,129 @@ impl<R: RngCore + CryptoRng> Planner<R> {
             balance: Balance::default(),
             vote_intents: BTreeMap::default(),
             plan: TransactionPlan::default(),
-            ibc_actions: Vec::new(),
             gas_prices: GasPrices::zero(),
             fee_tier: FeeTier::default(),
+            actions: Vec::new(),
+            change_outputs: BTreeMap::new(),
         }
     }
 
+    fn calculate_balance(&self) -> Balance {
+        let mut balance = Balance::zero();
+        for action in &self.actions {
+            balance += action.balance();
+        }
+        for action in self.change_outputs.values() {
+            balance += action.balance();
+        }
+
+        balance
+    }
+
+    fn calculate_balance_with_fees(&self, base_fee_estimation: Fee) -> Balance {
+        let mut balance = Balance::zero();
+        for action in &self.actions {
+            balance += action.balance();
+        }
+
+        for action in self.change_outputs.values() {
+            balance += action.balance();
+        }
+
+        balance -= base_fee_estimation.0;
+
+        balance
+    }
+
+    fn push(&mut self, action: ActionPlan) {
+        self.actions.push(action);
+    }
+
+    fn gas_estimate(&self) -> Gas {
+        // TODO: this won't include the gas cost for the bytes of the tx itself
+        // so this gas estimate will be an underestimate, but since the tx-bytes contribution
+        // to the fee is ideally small, hopefully it doesn't matter.
+        let mut gas = Gas::zero();
+        for action in &self.actions {
+            // TODO missing AddAssign
+            gas = gas + action.gas_cost();
+        }
+        for action in self.change_outputs.values() {
+            // TODO missing AddAssign
+            // TODO missing GasCost impl on OutputPlan
```

Comment:
In the original implementation, none of the `*Plan` structs had `GasCost` impls, but that can probably change

Reply:
adding the `GasCost` impls